### PR TITLE
On upgrade room only send canonical alias once.

### PR DESCRIPTION
**Pull request by erikjohnston**
Instead of doing a complicated dance of deleting and moving aliases one
by one, which sends a canonical alias update into the old room for each
one, lets do it all in one go.

This also changes the function to move *all* local alias events to the new
room, however that happens later on anyway.

### CODE REVIEW

**Comment by erikjohnston** [comment]
> Looks fine from a python and room upgrade perspective, but I'd like to get someone more familiar with the new alias semantics to review as well.

I believe this is keeping existing behaviour, we have some pretty detailed sytests for this

File: synapse/handlers/room.py [hunk 1]
```diff
@@ -439,73 +439,78 @@ async def _move_aliases_to_new_room(
...
-                # I'm not really expecting this to happen, but it could if the spam
-                # checking module decides it shouldn't, or similar.
-                logger.error("Error adding alias %s to new room: %s", alias, e)
+        # If there is a canonical alias we need to update the one in the old
+        # room and set one in the new one.
+        old_canonical_alias_content = dict(canonical_alias_event.content)
+        new_canonical_alias_content = {}
+
+        canonical = canonical_alias_event.content.get("alias")
+        if canonical and self.hs.is_mine_id(canonical):
```
Comment (anoadragon453) [inline synapse/handlers/room.py:459]: Are we not able to move canonical aliases if they don't belong to us? It's kind of annoying if the person upgrading a room (with canonical alias `#project:matrix.org` for visibility) has to move it via their matrix.org account.

Comment (erikjohnston) [reply synapse/handlers/room.py:459]: So we check the canonical alias event to make sure that all the aliases listed do actually point to the room. We can't a) move aliases that don't belong to the server, nor b) set the canonical alias if its still pointing to the old room (and I'm not sure we'd want to).

Comment (anoadragon453) [reply synapse/handlers/room.py:459]: Ah, yes of course. So in this case the room will be upgraded, and the alias will be left behind until someone from the server that owns it, moves it.

That actually does sound intuitive.

**Pull request merged**